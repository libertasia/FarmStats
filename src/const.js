export const AppRoute = {
  ROOT: `/`,
};

export const APIRoute = {
  FARMS: `/v1/farms`,
};

export const TabTypes = {
  GRAPHS: `graphs`,
  TABLE: `table`,
};

export const TableHeaders = [
  {
    title: `Location`,
  },
  {
    title: `Datetime`,
  },
  {
    title: `Metric`,
  },
  {
    title: `Value`,
  }
];
