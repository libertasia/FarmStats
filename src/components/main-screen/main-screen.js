import React, {useEffect} from "react";
import PropTypes from "prop-types";
import {connect} from "react-redux";
import Header from "../header/header";
import Footer from "../footer/footer";
import Tabs from "../tabs/tabs";
import DataTable from "../data-table/data-table";
import Filter from "../filter/filter";
import DataGraphs from "../data-graphs/data-graphs";
import {getActiveTab, getAllFarmsLoadedStatus, getAllFarmsStatsLoadedStatus, getFilteredFarmsStats, getAllFarmsStatsLoadingStatus} from "../../store/selectors";
import {fetchAllFarmsStats, fetchMonthlyStats} from "../../store/api-actions";
import {FarmsStatsShape, TabType} from "../../const";

const MainScreen = (props) => {
  const {activeTab, isAllFarmsStatsLoaded, isAllFarmsLoaded, loadFarmsStats, filteredStats} = props;

  useEffect(() => {
    if (isAllFarmsLoaded && !isAllFarmsStatsLoaded) {
      loadFarmsStats();
    }
  }, [isAllFarmsStatsLoaded, isAllFarmsLoaded]);

  return (
    <React.Fragment>
      <Header />
      <main className="wrapper__main container">
        <h1 className="visually-hidden">Solita Farms Statistics</h1>
        <Filter />
        <Tabs />
        {activeTab === TabType.TABLE &&
          <DataTable
            data={filteredStats}
          />
        }
        {activeTab === TabType.GRAPHS &&
          <DataGraphs />
        }
      </main>
      <Footer />
    </React.Fragment>
  );
};

MainScreen.propTypes = {
  activeTab: PropTypes.string.isRequired,
  isAllFarmsStatsLoaded: PropTypes.bool.isRequired,
  isAllFarmsLoaded: PropTypes.bool.isRequired,
  loadFarmsStats: PropTypes.func.isRequired,
  filteredStats: FarmsStatsShape,
};

const mapStateToProps = (state) => ({
  activeTab: getActiveTab(state),
  filteredStats: getFilteredFarmsStats(state),
  isAllFarmsStatsLoaded: getAllFarmsStatsLoadedStatus(state),
  isAllFarmStatsLoading: getAllFarmsStatsLoadingStatus(state),
  isAllFarmsLoaded: getAllFarmsLoadedStatus(state),
});

const mapDispatchToProps = (dispatch) => ({
  loadFarmsStats() {
    dispatch(fetchAllFarmsStats());
    dispatch(fetchMonthlyStats());
  },
});

export {MainScreen};
export default connect(mapStateToProps, mapDispatchToProps)(MainScreen);
