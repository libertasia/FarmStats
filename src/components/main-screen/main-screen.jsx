import React, {useEffect} from "react";
import {connect} from "react-redux";
import PropTypes from "prop-types";
import Header from "../header/header";
import Footer from "../footer/footer";
import Tabs from "../tabs/tabs";
import DataTable from "../data-table/data-table";
import {getActiveTab, getAllFarmsLoadedStatus, getAllFarmsStatsLoadedStatus, getFilteredFarmsStats, getAllFarmsStatsLoadingStatus} from "../../store/selectors";
import {fetchAllFarmsStats} from "../../store/api-actions";
import {TabType} from "../../const";
import Filter from "../filter/filter";

const MainScreen = (props) => {
  const {activeTab, isAllFarmsStatsLoaded, isAllFarmStatsLoading, isAllFarmsLoaded, loadFarmsStats, filteredStats} = props;

  useEffect(() => {
    if (isAllFarmsLoaded && !isAllFarmStatsLoading && !isAllFarmsStatsLoaded) {
      loadFarmsStats();
    }
  }, [isAllFarmsStatsLoaded, isAllFarmsLoaded, isAllFarmStatsLoading]);

  return (
    <React.Fragment>
      <Header />
      <main className="wrapper__main container">
        <h1 className="visually-hidden">Solita Farms Statistics</h1>
        <Tabs />
        {activeTab === TabType.TABLE &&
          <section>
            <Filter />
            <DataTable
              data={filteredStats}
            />
          </section>
        }
      </main>
      <Footer />
    </React.Fragment>
  );
};

MainScreen.propTypes = {
  allFarmsStats: PropTypes.array,
};

const mapStateToProps = (state) => ({
  activeTab: getActiveTab(state),
  filteredStats: getFilteredFarmsStats(state),
  isAllFarmsStatsLoaded: getAllFarmsStatsLoadedStatus(state),
  isAllFarmStatsLoading: getAllFarmsStatsLoadingStatus(state),
  isAllFarmsLoaded: getAllFarmsLoadedStatus(state),
});

const mapDispatchToProps = (dispatch) => ({
  loadFarmsStats() {
    dispatch(fetchAllFarmsStats());
  },
});

export {MainScreen};
export default connect(mapStateToProps, mapDispatchToProps)(MainScreen);
