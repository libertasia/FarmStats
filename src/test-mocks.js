import {MetricType} from "./const";

export const mockStats = [
  {
    "location": "farm 1",
    "datetime": "2018-12-31T22:00:00.000Z",
    "value": 5.88,
    "farm_id": "1",
    "sensor_type": "ph"
  },
  {
    "location": "farm 2",
    "datetime": "2018-12-31T22:00:00.000Z",
    "value": 5.88,
    "farm_id": "2",
    "sensor_type": "rainfall"
  },
  {
    "location": "farm 3",
    "datetime": "2018-12-31T22:00:00.000Z",
    "value": 5.88,
    "farm_id": "3",
    "sensor_type": "rainfall"
  },
  {
    "location": "farm 4",
    "datetime": "2018-12-31T22:00:00.000Z",
    "value": 5.88,
    "farm_id": "4",
    "sensor_type": "rainfall"
  },
];

export const mockLocations = [
  {
    "farm_id": "1",
    "name": "farm 1",
    "isSelected": true,
  },
  {
    "farm_id": "2",
    "name": "farm 2",
    "isSelected": true,
  },
  {
    "farm_id": "3",
    "name": "farm 3",
    "isSelected": true,
  },
  {
    "farm_id": "4",
    "name": "farm 4",
    "isSelected": true,
  }
];

export const mockMetrics = [
  {
    type: MetricType.PH,
    isSelected: true,
  },
  {
    type: MetricType.RAINFALL,
    isSelected: true,
  },
  {
    type: MetricType.TEMPERATURE,
    isSelected: true,
  }
];

export const mockMonthlyStats = [
  {
    "farm_id": "1",
    "sensor_type": "temperature",
    "stats": [
      {
        "month": 12,
        "year": 2019,
        "average": -5.3416,
        "median": -5.8,
        "standard_deviation": 5.0244
      },
      {
        "month": 1,
        "year": 2021,
        "average": -6.228,
        "median": -6,
        "standard_deviation": 4.9634
      },
      {
        "month": 9,
        "year": 2019,
        "average": 21.366,
        "median": 22.5,
        "standard_deviation": 4.5688
      },
    ]
  },
  {
    "farm_id": "2",
    "sensor_type": "temperature",
    "stats": [
      {
        "month": 12,
        "year": 2019,
        "average": -5.3416,
        "median": -5.8,
        "standard_deviation": 5.0244
      },
      {
        "month": 1,
        "year": 2021,
        "average": -6.228,
        "median": -6,
        "standard_deviation": 4.9634
      },
      {
        "month": 9,
        "year": 2019,
        "average": 21.366,
        "median": 22.5,
        "standard_deviation": 4.5688
      },
    ]
  },
  {
    "farm_id": "3",
    "sensor_type": "temperature",
    "stats": [
      {
        "month": 12,
        "year": 2019,
        "average": -5.3416,
        "median": -5.8,
        "standard_deviation": 5.0244
      },
      {
        "month": 1,
        "year": 2021,
        "average": -6.228,
        "median": -6,
        "standard_deviation": 4.9634
      },
      {
        "month": 9,
        "year": 2019,
        "average": 21.366,
        "median": 22.5,
        "standard_deviation": 4.5688
      },
    ]
  },
  {
    "farm_id": "4",
    "sensor_type": "temperature",
    "stats": [
      {
        "month": 12,
        "year": 2019,
        "average": -5.3416,
        "median": -5.8,
        "standard_deviation": 5.0244
      },
      {
        "month": 1,
        "year": 2021,
        "average": -6.228,
        "median": -6,
        "standard_deviation": 4.9634
      },
      {
        "month": 9,
        "year": 2019,
        "average": 21.366,
        "median": 22.5,
        "standard_deviation": 4.5688
      },
    ]
  },
];

export const mockFarms = [
  {
    "farm_id": "1",
    "name": "farm 1",
    "location": "Metsola"
  },
  {
    "farm_id": "2",
    "name": "farm 2",
    "location": "Viikki"
  },
  {
    "farm_id": "3",
    "name": "farm 3",
    "location": "Tampere"
  },
  {
    "farm_id": "4",
    "name": "farm 4",
    "location": "Etel√§esplanadi"
  },
];
