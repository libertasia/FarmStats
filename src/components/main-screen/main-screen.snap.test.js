import React from "react";
import {Provider} from "react-redux";
import {render} from "@testing-library/react";
import configureStore from "redux-mock-store";
import {TabType} from "../../const";
import {MainScreen} from "./main-screen";
import {mockLocations, mockMetrics, mockStats} from "../../test-mocks";

const mockStore = configureStore({});
const store = {
  DATA: {
    farmsStats: mockStats,
  },
  VIEW: {
    activeTab: TabType.TABLE,
    locations: mockLocations,
    metrics: mockMetrics,
  }
};

it(`MainScreen renders correctly`, () => {
  const {container} = render(
      <Provider store={mockStore(store)}>
        <MainScreen
          activeTab={TabType.TABLE}
          isAllFarmsStatsLoaded={true}
          isAllFarmsLoaded={true}
          loadFarmsStats={jest.fn()}
          filteredStats={mockStats}
        />
      </Provider>
  );

  expect(container).toMatchSnapshot();
});
