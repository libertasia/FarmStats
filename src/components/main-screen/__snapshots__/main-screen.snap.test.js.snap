// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MainScreen renders correctly 1`] = `
<div>
  <header
    class="wrapper__header header"
  >
    <div
      class="header__wrapper container"
    >
      <p>
        Solita Farms Statistics
      </p>
    </div>
  </header>
  <main
    class="wrapper__main container"
  >
    <h1
      class="visually-hidden"
    >
      Solita Farms Statistics
    </h1>
    <section
      class="filter"
    >
      <h2>
        Filter options
      </h2>
      <form
        action="#"
        class="filter__form"
        method="get"
      >
        <section
          class="filter__form-section"
        >
          <h3>
            Location
          </h3>
          <ul
            class="filter__list"
          >
            <li
              class="filter__list-item"
            >
              <input
                checked=""
                id="farm 1"
                name="1"
                type="checkbox"
              />
              <label
                for="farm 1"
              >
                farm 1
              </label>
            </li>
            <li
              class="filter__list-item"
            >
              <input
                checked=""
                id="farm 2"
                name="2"
                type="checkbox"
              />
              <label
                for="farm 2"
              >
                farm 2
              </label>
            </li>
            <li
              class="filter__list-item"
            >
              <input
                checked=""
                id="farm 3"
                name="3"
                type="checkbox"
              />
              <label
                for="farm 3"
              >
                farm 3
              </label>
            </li>
            <li
              class="filter__list-item"
            >
              <input
                checked=""
                id="farm 4"
                name="4"
                type="checkbox"
              />
              <label
                for="farm 4"
              >
                farm 4
              </label>
            </li>
          </ul>
        </section>
        <section
          class="filter__form-section"
        >
          <h3>
            Metric
          </h3>
          <ul
            class="filter__list"
          >
            <li
              class="filter__list-item"
            >
              <input
                checked=""
                id="ph"
                name="ph"
                type="checkbox"
              />
              <label
                for="ph"
              >
                ph
              </label>
            </li>
            <li
              class="filter__list-item"
            >
              <input
                checked=""
                id="rainfall"
                name="rainfall"
                type="checkbox"
              />
              <label
                for="rainfall"
              >
                rainfall
              </label>
            </li>
            <li
              class="filter__list-item"
            >
              <input
                checked=""
                id="temperature"
                name="temperature"
                type="checkbox"
              />
              <label
                for="temperature"
              >
                temperature
              </label>
            </li>
          </ul>
        </section>
      </form>
    </section>
    <div
      class="tabs"
    >
      <ul
        class="tabs__list"
      >
        <li
          class="tabs__list-item tabs__list-item--active"
          data-id="table"
        >
          <button
            class="tabs__link-btn"
            type="button"
          >
            table
          </button>
        </li>
        <li
          class="tabs__list-item tabs__list-item--inactive"
          data-id="graphs"
        >
          <button
            class="tabs__link-btn"
            type="button"
          >
            graphs
          </button>
        </li>
      </ul>
    </div>
    <div
      class="table-wrapper"
    >
      <table
        class="data-table"
        role="table"
      >
        <caption
          class="visually-hidden"
        >
          Farms Data
        </caption>
        <thead>
          <tr>
            <th>
              Location
            </th>
            <th>
              Datetime (UTC)
            </th>
            <th>
              Metric
            </th>
            <th>
              Value
            </th>
          </tr>
        </thead>
        <tbody>
          <tr>
            <td>
              farm 1
            </td>
            <td>
              <time
                datetime="2018-12-31T22:00:00.000Z"
              >
                2018-12-31, 22:00
              </time>
            </td>
            <td>
              ph
            </td>
            <td>
              5.88
            </td>
          </tr>
          <tr>
            <td>
              farm 2
            </td>
            <td>
              <time
                datetime="2018-12-31T22:00:00.000Z"
              >
                2018-12-31, 22:00
              </time>
            </td>
            <td>
              rainfall
            </td>
            <td>
              5.88
            </td>
          </tr>
          <tr>
            <td>
              farm 3
            </td>
            <td>
              <time
                datetime="2018-12-31T22:00:00.000Z"
              >
                2018-12-31, 22:00
              </time>
            </td>
            <td>
              rainfall
            </td>
            <td>
              5.88
            </td>
          </tr>
          <tr>
            <td>
              farm 4
            </td>
            <td>
              <time
                datetime="2018-12-31T22:00:00.000Z"
              >
                2018-12-31, 22:00
              </time>
            </td>
            <td>
              rainfall
            </td>
            <td>
              5.88
            </td>
          </tr>
        </tbody>
      </table>
      <div
        class="MuiTablePagination-root css-jtlhu6-MuiTablePagination-root"
      >
        <div
          class="MuiToolbar-root MuiToolbar-gutters MuiToolbar-regular MuiTablePagination-toolbar css-11bfvty-MuiToolbar-root-MuiTablePagination-toolbar"
        >
          <div
            class="MuiTablePagination-spacer css-1psng7p-MuiTablePagination-spacer"
          />
          <p
            class="MuiTablePagination-selectLabel css-pdct74-MuiTablePagination-selectLabel"
            id="mui-2"
          >
            Rows per page:
          </p>
          <div
            class="MuiInputBase-root MuiInputBase-colorPrimary css-16c50h-MuiInputBase-root-MuiTablePagination-select"
          >
            <div
              aria-expanded="false"
              aria-haspopup="listbox"
              aria-labelledby="mui-2 mui-1"
              class="MuiTablePagination-select MuiSelect-select MuiSelect-standard MuiInputBase-input css-194a1fa-MuiSelect-select-MuiInputBase-input"
              id="mui-1"
              role="button"
              tabindex="0"
            >
              10
            </div>
            <input
              aria-hidden="true"
              class="MuiSelect-nativeInput css-yf8vq0-MuiSelect-nativeInput"
              tabindex="-1"
              value="10"
            />
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiTablePagination-selectIcon MuiSelect-icon MuiSelect-iconStandard css-pqjvzy-MuiSvgIcon-root-MuiSelect-icon"
              data-testid="ArrowDropDownIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M7 10l5 5 5-5z"
              />
            </svg>
          </div>
          <p
            class="MuiTablePagination-displayedRows css-levciy-MuiTablePagination-displayedRows"
          >
            1â€“4 of 4
          </p>
          <div
            class="MuiTablePagination-actions"
          >
            <button
              aria-label="Go to previous page"
              class="MuiButtonBase-root Mui-disabled MuiIconButton-root Mui-disabled MuiIconButton-colorInherit MuiIconButton-sizeMedium css-zylse7-MuiButtonBase-root-MuiIconButton-root"
              disabled=""
              tabindex="-1"
              title="Go to previous page"
              type="button"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                data-testid="KeyboardArrowLeftIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M15.41 16.09l-4.58-4.59 4.58-4.59L14 5.5l-6 6 6 6z"
                />
              </svg>
            </button>
            <button
              aria-label="Go to next page"
              class="MuiButtonBase-root Mui-disabled MuiIconButton-root Mui-disabled MuiIconButton-colorInherit MuiIconButton-sizeMedium css-zylse7-MuiButtonBase-root-MuiIconButton-root"
              disabled=""
              tabindex="-1"
              title="Go to next page"
              type="button"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                data-testid="KeyboardArrowRightIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M8.59 16.34l4.58-4.59-4.58-4.59L10 5.75l6 6-6 6z"
                />
              </svg>
            </button>
          </div>
        </div>
      </div>
    </div>
  </main>
  <footer
    class="wrapper__footer footer"
  >
    <div
      class="footer__wrapper container"
    >
      <p>
        Copyright Â© 2021 Anastasiia Erokhina. All rights reserved.
      </p>
    </div>
  </footer>
</div>
`;
